dali/python/nvidia/dali/pickling.py

```diff
@@ -0,0 +1,144 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import pickle
+import sys
+import types
+import marshal
+import importlib
+import io
+
+dummy_lambda = lambda : 0
+
+def set_funcion_state(fun, state):
+    fun.__globals__.update(state['global_refs'])
+    fun.__defaults__ = state['defaults']
+    fun.__kwdefaults__ = state['kwdefaults']
+
+def function_unpickle(name, qualname, code, closure):
+    code = marshal.loads(code)
+    globs = {'__builtins__': __builtins__}
+    fun = types.FunctionType(code, globs, name, closure=closure)
+    fun.__qualname__ = qualname
+    return fun
+
+def function_by_value_reducer(fun):
+    cl_vars = inspect.getclosurevars(fun)
+    code = marshal.dumps(fun.__code__)
+    basic_def = (fun.__name__, fun.__qualname__, code, fun.__closure__)
+    fun_context = {
+        'global_refs': cl_vars.globals,
+        'defaults': fun.__defaults__,
+        'kwdefaults': fun.__kwdefaults__
+    }
+    return function_unpickle, basic_def, fun_context, None, None, set_funcion_state
+
+def module_unpickle(name, origin, submodule_search_locations):
+    if name in sys.modules:
+        return sys.modules[name]
+    spec = importlib.util.spec_from_file_location(
+        name, origin,
+        submodule_search_locations=submodule_search_locations)
+    module = importlib.util.module_from_spec(spec)
+    sys.modules[name] = module
+    spec.loader.exec_module(module)
+    return module
+
+def module_reducer(module):
+    spec = module.__spec__
+    return module_unpickle, (spec.name, spec.origin, spec.submodule_search_locations)
+
+def set_cell_state(cell, state):
+    cell.cell_contents = state['cell_contents']
+
+def cell_unpickle():
+    return types.CellType(None)
+
+def cell_reducer(cell):
+    return (cell_unpickle, tuple(), {'cell_contents': cell.cell_contents}, None, None, set_cell_state)
+
+
+class DaliCallbackPickler(pickle.Pickler):
+
+    def reducer_override(self, obj):
+        if inspect.ismodule(obj):
+            return module_reducer(obj)
+        if isinstance(obj, types.CellType):
+            return cell_reducer(obj)
+        if inspect.isfunction(obj):
+            if isinstance(obj, type(dummy_lambda)) and obj.__name__ == dummy_lambda.__name__ or \
+                    getattr(obj, '_dali_pickle_by_value', False):
+                return function_by_value_reducer(obj)
+            if '<locals>' in obj.__qualname__:
```

Comment:
Does it work with regular local functions, such as:
```
def f():
  def g(): return 42;
```
?
`g.__qualname__ == 'f.<locals>.g'`, but it's otherwise quite trivial.
How about switching the logic to:
```Python:
try:
  pickle.dumps(obj)
except (whatever it raises):
  return function_by_value_reducer_(obj)
```
?

Reply:
Yep, g will be serialized by value and plain pickling fails as pickle recursively steps down the attributes and raises AttributeError when encountering <locals>.

I'd rather limit this `pickle.dumps` test to the situations when I am nearly sure that it is going to fail (or maybe should get rid of the test whatsoever?) for efficiency reasons. If the test succeeded it serialized something only for us to discard that result and proceed with a single step of serialization.

It would be nice if we could catch the exception by overriding dump method of the pickler, catch AttributeError and only then fall back to obj.__dali_pickle_by_value = True. Unfortunately it doesn't seem to be (easily) doable as the pickle.Pickler class is in fact a wrapper around C implementation of pickle and overridden dump method would be called only once for a top level object; recursive steps for dependencies/contents will be serialized by C code unaware of override.